packages/bloc_test/lib/src/bloc_test.dart

```diff
@@ -170,6 +171,271 @@ void blocTest<B extends BlocBase<State>, State>(
+///   act: (bloc, fakeAsync) {
+///     bloc
+///       ..add(CounterEvent.increment)
+///       ..add(CounterEvent.increment);
+///   },
+///   skip: 1,
+///   expect: () => [2],
+/// );
+/// ```
+///
+/// [fakeAsyncBlocTest] can also be used to wait for async operations
+/// by optionally providing a `Duration` to [wait].
+///
+/// ```dart
+/// blocTest(
+///   'CounterBloc emits [1] when increment is added',
+///   build: () => CounterBloc(),
+///   act: (bloc) => bloc.add(CounterEvent.increment),
+///   wait: const Duration(milliseconds: 300),
+///   expect: () => [1],
+/// );
+/// ```
+///
+/// [fakeAsyncBlocTest] can also be used to [verify] internal bloc
+/// functionality.
+///
+/// ```dart
+/// blocTest(
+///   'CounterBloc emits [1] when increment is added',
+///   build: () => CounterBloc(),
+///   act: (bloc, fakeAsync) => bloc.add(CounterEvent.increment),
+///   expect: () => [1],
+///   verify: (_) {
+///     verify(() => repository.someMethod(any())).called(1);
+///   }
+/// );
+/// ```
+///
+/// **Note:** when using [fakeAsyncBlocTest] with state classes which don't
+/// override `==` and `hashCode` you can provide an `Iterable` of matchers
+/// instead of explicit state instances.
+///
+/// ```dart
+/// blocTest(
+///  'emits [StateB] when EventB is added',
+///  build: () => MyBloc(),
+///  act: (bloc, fakeAsync) => bloc.add(EventB()),
+///  expect: () => [isA<StateB>()],
+/// );
+/// ```
+///
+/// If [tags] is passed, it declares user-defined tags that are applied to the
+/// test. These tags can be used to select or skip the test on the command line,
+/// or to do bulk test configuration. All tags should be declared in the
+/// [package configuration file][configuring tags]. The parameter can be an
+/// [Iterable] of tag names, or a [String] representing a single tag.
+///
+/// [configuring tags]: https://github.com/dart-lang/test/blob/master/pkgs/test/doc/configuration.md#configuring-tags
+@isTest
+void fakeAsyncBlocTest<B extends BlocBase<State>, State>(
+  String description, {
+  void Function(FakeAsync async)? setUp,
+  required B Function() build,
+  State Function()? seed,
+  Function(B bloc, FakeAsync async)? act,
+  Duration? wait,
+  int skip = 0,
+  dynamic Function()? expect,
+  Function(B bloc)? verify,
+  dynamic Function()? errors,
+  void Function(FakeAsync async)? tearDown,
+  dynamic tags,
+}) {
+  test.test(
+    description,
+    () {
+      testBlocFakeAsync<B, State>(
+        setUp: setUp,
+        build: build,
+        seed: seed,
+        act: act,
+        wait: wait,
+        skip: skip,
+        expect: expect,
+        verify: verify,
+        errors: errors,
+        tearDown: tearDown,
+      );
+    },
+    tags: tags,
+  );
+}
+
+/// Internal [testBlocFakeAsync] runner which is only visible for testing.
+/// This should never be used directly -- please use [fakeAsyncBlocTest]
+/// instead.
+@visibleForTesting
+void testBlocFakeAsync<B extends BlocBase<State>, State>({
```

Comment:
This method duplicates a lot of code that can be found on `testBloc`. What about refactoring both to share as much code as possible so maintenance is simpler?

Reply:
@tenhobi @mugbug there's a few reasons to not unify the code: 

1. Breaking change
2. **fakeAsync** require sync execution flow, which lead us to critical differences as: 
- runZonedGuarded sync execution
- different error handling
- possible confusion / false positive results of usage same method for fakeAsync / normal execution - https://github.com/dart-lang/fake_async/issues/38
- and most important fakeAsync issues, which prevent us to mix fakeAsync / normal execution in the same Zone: 
     1. https://github.com/dart-lang/fake_async/issues/51
     2. https://github.com/dart-lang/fake_async/issues/24

Reply:
Got your point, @mugbug  The duplicated code has been extracted to a private functions.

cc: @tenhobi

Reply:
LGTM! Now I suppose we just need to wait for an approval on the code solution and then work on some examples and documentation updates. Good job!